### PR TITLE
Add debug strategy plugin

### PR DESCRIPTION
**Pull request by ks888**
##### ISSUE TYPE
- Feature Pull Request
##### ANSIBLE VERSION

```
ansible 2.1.0 (add-debug-strategy-plugin db83400d71) last updated 2016/03/24 01:47:52 (GMT +900)
  lib/ansible/modules/core: (detached HEAD 7efc09ef08) last updated 2016/03/24 01:12:49 (GMT +900)
  lib/ansible/modules/extras: (detached HEAD 7f9cdc0350) last updated 2016/03/24 01:13:10 (GMT +900)
  config file =
  configured module search path = Default w/o overrides
```
##### SUMMARY

Add "debug" strategy. It enables you to invoke a debugger when a task is failed, and check several info, such as the value of a variable. Also, it is possible to update module arguments in the debugger, and run the failed task again with new arguments to consider how you can fix an issue.

There is a post about this strategy:
https://groups.google.com/forum/#!topic/ansible-project/IzjDLiIORlw

### CODE REVIEW

**Comment by sivel** [comment]
This should probably have some associated documentation included with it.

**Comment by bcoca** [comment]
We can get docs from the readme https://github.com/ks888/ansible-playbook-debugger

**Comment by ks888** [comment]
I've added usage document. It's a part of "Playbooks: Special Topics" and almost same as the original readme https://github.com/ks888/ansible-playbook-debugger, but written in reST.

**Comment by ks888** [comment]
@sivel @bcoca I've added usage document. It's a part of "Playbooks: Special Topics" and almost same as the original readme https://github.com/ks888/ansible-playbook-debugger, but written in reST.

**Comment by jpic** [comment]
Awesome !!! I just wanted to let you know it works for me:

```
$ cat test.yml 
- hosts: localhost
  connection: local
  strategy: debug
  gather_facts: no
  vars:
    var1: value1
  tasks:
    - name: wrong variable
      ping: data={{ wrong_var }}

$ ansible-playbook -i localhost, test.yml
PLAY [localhost] ***************************************************************

TASK [wrong variable] **********************************************************
fatal: [localhost]: FAILED! => {"failed": true, "msg": "'wrong_var' is undefined"}
Debugger invoked
(debug) p result
{'failed': True, 'msg': u"'wrong_var' is undefined"}
(debug) p task.args
{u'data': u'{{ wrong_var }}'}
(debug) redo
fatal: [localhost]: FAILED! => {"failed": true, "msg": "'wrong_var' is undefined"}
Debugger invoked
(debug) task.args['data'] = '{{ var1 }}'
(debug) redo
ok: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0   
```

Perhaps some tests would be nice to have.

**Comment by jpic** [comment]
Not sure why, but in a real situation context I had to do this change:

```

+++ b/lib/ansible/plugins/strategy/debug.py
@@ -6,7 +6,8 @@ import cmd
 import pprint
 import sys

-from ansible.plugins.strategy import linear
+from ansible.plugins.strategy.linear import StrategyModule as LinearStrategyModule
 from ansible.plugins.strategy import StrategyBase

 try:
@@ -26,7 +27,7 @@ class NextAction(object):
         self.result = result


-class StrategyModule(linear.StrategyModule, StrategyBase):
+class StrategyModule(LinearStrategyModule, StrategyBase):
     # Usually inheriting linear.StrategyModule is enough. However, StrategyBase class must be
     # direct ancestor to be considered as strategy plugin, and so we inherit the class here.
```

To avoid:

```

the full traceback was:

Traceback (most recent call last):
  File "/home/jpic/env/src/ansible/bin/ansible-playbook", line 86, in <module>
    sys.exit(cli.run())
  File "/home/jpic/env/src/ansible/lib/ansible/cli/playbook.py", line 151, in run
    results = pbex.run()
  File "/home/jpic/env/src/ansible/lib/ansible/executor/playbook_executor.py", line 141, in run
    result = self._tqm.run(play=play)
  File "/home/jpic/env/src/ansible/lib/ansible/executor/task_queue_manager.py", line 225, in run
    strategy = strategy_loader.get(new_play.strategy, self)
  File "/home/jpic/env/src/ansible/lib/ansible/plugins/__init__.py", line 332, in get
    self._module_cache[path] = self._load_module_source('.'.join([self.package, name]), path)
  File "/home/jpic/env/src/ansible/lib/ansible/plugins/__init__.py", line 318, in _load_module_source
    module = imp.load_source(name, path, module_file)
  File "/home/jpic/env/src/ansible/lib/ansible/plugins/strategy/debug.py", line 9, in <module>
    from ansible.plugins.strategy import linear
ImportError: cannot import name linear
```

File: lib/ansible/plugins/strategy/debug.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    """ The next action after an interpreter's exit. """
+    REDO = 1
+    CONTINUE = 2
+    EXIT = 3
+
+    def __init__(self, result=EXIT):
+        self.result = result
+
+
+class StrategyModule(linear.StrategyModule, StrategyBase):
```
Comment (abadger) [inline lib/ansible/plugins/strategy/debug.py:29]: I think I recently fixed the bug which made you have to inherit directly from StrategyBase.  Try just leaving in linear.StrategyModule and see if it now works.

**Pull request merged**